pkg/cosign/tsa.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2024 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cosign
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"fmt"
+	"os"
+
+	"github.com/sigstore/cosign/v2/pkg/cosign/env"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"github.com/sigstore/sigstore/pkg/tuf"
+)
+
+const (
+	tsaLeafCertStr                = `tsa_leaf.crt.pem`
+	tsaRootCertStr                = `tsa_root.crt.pem`
+	tsaIntermediateCertStrPattern = `tsa_intermediate_%d.crt.pem`
+)
+
+type TSACertificates struct {
+	LeafCert          *x509.Certificate
+	IntermediateCerts []*x509.Certificate
+	RootCert          []*x509.Certificate
+}
+
+type GetTargetStub func(ctx context.Context, usage tuf.UsageKind, names []string) ([]*x509.Certificate, error)
+
+func GetTufTargets(ctx context.Context, usage tuf.UsageKind, names []string) ([]*x509.Certificate, error) {
+	tufClient, err := tuf.NewFromEnv(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error creating TUF client: %w", err)
+	}
+	var allCerts []*x509.Certificate
+	targets, err := tufClient.GetTargetsByMeta(usage, names)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching targets by metadata with usage %v: %w", usage, err)
+	}
+	for _, target := range targets {
+		certs, _, _, err := splitPEMCertificateChain(target.Target)
```

Comment:
I think this logic can be simplified. It looks like you're gathering all targets, unmarshalling each into an object, returning them, marshalling them back into PEM, then unmarshalling then.

Can we instead change this function to return a string which is the concatenation of all targets, split by newlines? This is equivalent to the `raw` variables.

You can remove lines 94-100. Then you'll call `splitPEMCertificateChain(raw)`, which you already do.